### PR TITLE
Introduce StringParser and StringNormalizer

### PR DESCRIPTION
**Pull request by brightbyte**
[Bug: T104874](http://phabricator.wikimedia.org/T104874)

### CODE REVIEW

File: src/ValueParsers/Normalizers/NullStringNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace ValueParsers\Normalizers;
+
+use InvalidArgumentException;
+
+/**
+ * Null implementation of StringNormalizer.
```
Comment (thiemowmde) [inline src/ValueParsers/Normalizers/NullStringNormalizer.php]: `@since`

File: src/ValueParsers/Normalizers/StringNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace ValueParsers\Normalizers;
+
+use InvalidArgumentException;
+
+/**
+ * Interface for string normalization
+ *
```
Comment (thiemowmde) [inline src/ValueParsers/Normalizers/StringNormalizer.php]: `@since`

File: src/ValueParsers/StringParser.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	/**
+	 * @var StringNormalizer
+	 */
+	private $normalizer;
+
+	/**
+	 * @param StringNormalizer $normalizer
+	 */
+	public function __construct( StringNormalizer $normalizer = null ) {
+		if ( $normalizer === null ) {
```
Comment (thiemowmde) [inline src/ValueParsers/StringParser.php]: I love the new `?:` syntax... :-)

File: src/ValueParsers/Normalizers/NullStringNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @license GPL 2+
+ * @author Daniel Kinzler
+ */
+class NullStringNormalizer implements StringNormalizer {
```
Comment (thiemowmde) [inline src/ValueParsers/Normalizers/NullStringNormalizer.php]: Oh noes, you forgot to write a test for this class. ;-)))

File: tests/ValueParsers/StringParserTest.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+namespace ValueParsers\Test;
+
+use DataValues\DataValue;
+use DataValues\StringValue;
+use ValueParsers\StringParser;
+
+/**
+ * Unit test StringParser class.
+ *
+ * @covers StringParser
```
Comment (thiemowmde) [inline tests/ValueParsers/StringParserTest.php]: No namespace?

File: src/ValueParsers/StringParser.php (hunk 3 above)
Comment (JeroenDeDauw) [inline src/ValueParsers/StringParser.php]: Having a single assignment would be better than having two. No need for this much state

File: tests/ValueParsers/StringParserTest.php (hunk 5 above)
Comment (JeroenDeDauw) [inline tests/ValueParsers/StringParserTest.php]: This will error out on ScrutinizerCI yes...

File: tests/ValueParsers/StringParserTest.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+			'null' => array( null ),
+			'array' => array( array() ),
+			'int' => array( 7 ),
+		);
+	}
+
+	/**
+	 * @dataProvider provideParse_InvalidArgumentException
+	 */
+	public function testParse_InvalidArgumentException( $input ) {
```
Comment (JeroenDeDauw) [inline tests/ValueParsers/StringParserTest.php]: `testFoo_bar`, not `testFoo_Bar`

`testGivenNonString_exceptionIsThrown` describes the expectation better

**Comment by JeroenDeDauw** [comment]
I thought we had something like this already... apparently not. Perhaps we removed it at some point?

**Comment by thiemowmde** [comment]
-1 because we do have a `StringValueParser` in the same directory, but it's abstract. This patch introduces an actual StringValueParser but calls it StringParser because the class name is already ocupied. I do find this confusing. Is there a way to fix this? We are targeting for a 1.0.0 release anyway.

**Comment by JeroenDeDauw** [comment]
@thiemowmde actually all concrete parsers in this directory do not have "value" in their name. They are named after what they parse, not after the value that represents the result of the parsing. There often are multiple ways to represent a given value, you can have multiple parsers that output the same one (eg the geocoord ones), so this makes a lot of sense.

**Comment by thiemowmde** [comment]
EntityIdValueParser being the only exception. Ok, removing my -1 for the name StringParser. Thanks for pointing that out.

However, this leaves the question why StringParser does not extend StringValueParser, such as all other string based parsers?

**Comment by brightbyte** [comment]
@thiemowmde StringParser  doesn't extend StringValueParser because it doesn't need anything StringValueParser provides. In fact, I think we should just kill StringValueParser.

**Comment by thiemowmde** [comment]
The base class does not bring relevant advantages, so I would almost vote +2, but would like to see @JeroenDeDauw's comment tackled.

**Comment by thiemowmde** [comment]
I renamed the test method. @brightbyte, please have a look at my commit 338f320184a513b35d3153e6cf147d8fda68e6e8 and feel free to merge. +2 from my side.

**Pull request merged**